### PR TITLE
Rewrite dynamic Groovy call interception to make it more generic

### PR DESCRIPTION
**Pull request by mlopatkin**
Groovy has a variety of ways to call methods/constructors or read
properties. One has to override a lot of methods of the CallSite to
properly intercept them: call() to intercept instance and static
methods, callStatic() to intercept calls to statically imported methods,
callGetProperty() to intercept property reads. There are "optimized"
calls to methods with separate parameters but there are also bulk
overloads that accept argument arrays. The latter are used for spread
calls (e.g. `foo(*[bar, baz])`).

This diversity caused our instrumentation to miss some cases, most
notably spread calls were not intercepted at all, and some calls weren't
interceped if static imports are used.

This commit separates the generic "interception" code that deals with
Groovy's CallSites and their arguments from the user code that performs
actions when the intercepted methods are called. The interceptor takes
care of overriding necessary methods, unwrapping the arguments, and
provides a way to call the original Groovy implementation if the user
code decides that the invocation shouldn't be intercepted.

The Invocation object hides the details of the actual Groovy machinery
involved in calling the method/constructor, so the interceptor can be
extended further to support Groovy's invokedynamic-based implementation
while reusing the user code as is.

One downside compared to the previous implementation is that optimized
code paths with separate arguments are no longer optimized as the
arguments are always packed into arrays before further processing.
It shouldn't be a big problem in practice as the "interesting" calls are
not that frequent and others are inlined by the Groovy runtime after the
first invocation. A proper microbenchmarking coverage is necessary to
optimize things further.

This is a first step of implementing #20794.

### CODE REVIEW

**Comment by mlopatkin** [comment]
@bot-gradle help

**Comment by mlopatkin** [comment]
@bot-gradle test APT, ACC

**Comment by mlopatkin** [comment]
Follow-up PR that adds indy instrumentation #20799.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java [hunk 1]
```diff
@@ -81,54 +82,36 @@ public static void discardListener() {
         setListener(NO_OP);
     }
 
+    private static final Map<String, CallInterceptor> GROOVY_CALLSITE_INTERCEPTORS = new HashMap<>();
+
+    static {
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java:87]: 💄 I'd use a static factory method here, `GROOVY_CALLSITE_INTERCEPTORS = buildCallSiteInterceptors();`

That way the factory method can use use a simpler name for the map as it builds it.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java [hunk 2]
```diff
@@ -81,54 +82,36 @@ public static void discardListener() {
...
-                case "start":
-                    array.array[callSite.getIndex()] = new ProcessBuilderStartCallSite(callSite);
-                    break;
-                case "startPipeline":
-                    array.array[callSite.getIndex()] = new ProcessBuilderStartPipelineCallSite(callSite);
-                    break;
-                case "<$constructor$>":
-                    array.array[callSite.getIndex()] = new FileInputStreamConstructorCallSite(callSite);
-                    break;
+            CallInterceptor interceptor = GROOVY_CALLSITE_INTERCEPTORS.get(callSite.getName());
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java:112]: 👍

File: subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+         * Throws {@link ArrayIndexOutOfBoundsException} if {@code pos} is outside the bounds.
+         *
+         * @param pos the position of the argument
+         * @return the unwrapped value of the argument
+         */
+        Object getArgument(int pos);
+
+        /**
+         * Returns an <b>unwrapped</b> argument at the position {@code pos} or {@code null} if the {@code pos} is greater or equal than {@link #getArgsCount()}.
+         * This method is useful for handling "telescopic" overloads.
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptor.java:62]: 🗒️ An example of a "telescopic" overload would be useful here.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/ClassBoundCallInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+public abstract class ClassBoundCallInterceptor extends CallInterceptor {
+    private final Class<?> expectedReceiver;
+
+    public ClassBoundCallInterceptor(Class<?> expectedReceiver) {
+        this.expectedReceiver = expectedReceiver;
+    }
+
+    @Override
+    protected final Object doIntercept(Invocation invocation, String consumer) throws Throwable {
+        if (!expectedReceiver.equals(invocation.getReceiver())) {
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/ClassBoundCallInterceptor.java:32]: ❓ Wouldn't it make sense to use reference inequality here?

File: subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java [hunk 5]
```diff
@@ -81,54 +82,36 @@ public static void discardListener() {
...
     }
 
+    private static final Map<String, CallInterceptor> GROOVY_CALLSITE_INTERCEPTORS = new HashMap<>();
+
+    static {
+        GROOVY_CALLSITE_INTERCEPTORS.put("getProperty", new SystemGetPropertyInterceptor());
+        GROOVY_CALLSITE_INTERCEPTORS.put("setProperty", new SystemSetPropertyInterceptor());
+
+        SystemGetPropertiesInterceptor getPropertiesInterceptor = new SystemGetPropertiesInterceptor();
+        GROOVY_CALLSITE_INTERCEPTORS.put("getProperties", getPropertiesInterceptor);
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java:92]: 💄 Perhaps move the list of applicable method names to each interceptor? It seems they are already tailored to specific methods anyway.

**Review by bamboo — APPROVED** [review]
I really appreciate the simplification and how the overload selection looks more precise.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptor.java (hunk 3 above)
Comment (mlopatkin) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptor.java]: Done

File: subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java (hunk 5 above)
Comment (mlopatkin) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java]: I've tried to do that, and ended up with a quite involved solution. However, I fixed some dispatch-related corner cases along the way and added the ability to intercept constructors of several classes, so the additional complexity might be justified. PTAL.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java (hunk 1 above)
Comment (mlopatkin) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/Instrumented.java]: Thank you for the suggestion! I solved it in another way.

**Comment by mlopatkin** [comment]
@bot-gradle test this please

**Comment by mlopatkin** [comment]
I did several runs with gradle-profiler and it seems that the performance difference is negligible even in pathological cases of running 1000 constructors or having 1000 `System.getenv` call sites in a loop.

**Comment by bamboo** [comment]
> I did several runs with gradle-profiler and it seems that the performance difference is negligible even in pathological cases of running 1000 constructors or having 1000 `System.getenv` call sites in a loop.

That's great, @mlopatkin 👍

**Comment by mlopatkin** [comment]
@bot-gradle test and merge

**Pull request merged**